### DISCUSSION TITLE
Serialize data class to Map<String, String>?

### DISCUSSION THREAD

**Discussion by vikasuy** [square/moshi#1583] [General]
Say I have the following data class

```
data class Test(val foo: String, val bar: Int)
```

With moshi, is there a way to serialize this to a  java or kotlin `Map<String, String>`?
e.g. i would want the result to be the equivalent of the following in kotlin:

```
val t = Test("foo-value", 42)
# should serialize to the following:
mapOf(
  'foo' -> "foo-value",
  "bar" -> 42
)
```

In a way this is one step above the result of moshi's `toJson` function

**Comment by JakeWharton** [comment]
Yes you'll want to use the [`toJsonValue`](https://square.github.io/moshi/1.x/moshi/moshi/com.squareup.moshi/-json-adapter/to-json-value.html) API which returns an `Any`.

In this case, since your type is a class, that will result in the value being a `Map<String, Any>`. And since your class contains only scalar values (i.e., its properties are not also objects themselves) the map values will also both be a `String`.

**Reply by vikasuy** [reply]
Ah good to know. Is there a way to get this to work with nested data classes as well?

```
data class Inner(val bar: Int)
data class Outer(val foo: String, val inner: Inner)
val test = Outer("foo-value", Inner(42))
# expected serialization
mapOf(
  "foo" -> "foo-value",
  "inner" -> "{\"bar\": 42}"
}

**Reply by JakeWharton** [reply]
Not out of the box. By default the nested object would be another Map.

You would have to write a custom adapter whose behavior was to serialize non-scalar properties to a string, apply it to your root object type, and then use that with toJsonValue.
